Add lazy closure evaluation to `default` (#14160)

Mrfiregem opened the pull request:
# Description

This PR adds lazy closure evaluation to the `default` command (closes #14160).

- For non-closure values and without providing a column name, `default` acts the same as before
- The user can now provide multiple column names to populate if empty
- If the user provides a column name, the input must be a record or list, otherwise an error is created.
- The user can now provide a closure as a default value
  - This closure is run without any arguments or input
  - The closure is never evaluated if the value isn't needed
  - Even when column names are supplied, the closure is only run once (and cached to prevent re-calling it)

For example:

```nushell
> default { 1 + 2 } # => 3
> null | default 3 a   # => previously `null`, now errors
> 1 | default { sleep 5sec; 3 } # => `1`, without waiting 5 seconds

> let optional_var = null; $optional_var | default { input 'Enter value: ' } # => Returns user input
> 5 | default { input 'Enter value: ' } # => `5`, without prompting user

> ls | default { sleep 5sec; 'N/A' } name # => No-op since `name` column is never empty
> ls | default { sleep 5sec; 'N/A' } foo bar # => creates columns `foo` and `bar`; only takes 5 seconds since closure result is cached

# Old behavior is the same
> [] | default 'foo' # => []
> [] | default --empty 'foo' # => 'foo'
> default 5 # => 5
```

# User-Facing Changes

- Users can add default values to multiple columns now.
- Users can now use closures as the default value passed to `default`.
- To return a closure, the user must wrap the closure they want to return inside another closure, which will be run (`default { $my_closure }`).

# Tests + Formatting

All tests pass.

# After Submitting

weirdan commented:
It's worth noting that even closures can be returned by lazy `default`:

```nushell
default -l { {ls} }
```

Mrfiregem commented:
I was considering making closure evaluation the default, but it appears there's something relying on default sending a closure object since I got a panic when trying to `cargo run`. I believe it's prompt related, but I didn't look into it, and figured people might rely on the same thing for their scripts.

132ikl commented:
wait, you got a panic? that's not good, could you push a second branch to test what's up with that?

Mrfiregem commented:
That was from a previous attempt a few weeks back so I could be misremembering, and it was just an error in that code. I've fiddled with the current PR implementation to see if I could find it, but everything seems to work, so I think it's a false alarm.

As an aside, it does seem like we can enable lazy closure evaluation by default, if nobody would mind a breaking change for user scripts.
```diff
diff --git a/crates/nu-command/src/filters/default.rs b/crates/nu-command/src/filters/default.rs
index a1ce21148..f02c9e507 100644
--- a/crates/nu-command/src/filters/default.rs
+++ b/crates/nu-command/src/filters/default.rs
@@ -30,14 +30,14 @@ impl Command for Default {
                 Some('e'),
             )
             .switch(
-                "lazy",
-                "if default value is a closure, evaluate it",
-                Some('l'),
+                "no-eval",
+                "if default value is a closure, pass it as-is",
+                Some('n'),
             )
             .switch(
-                "lazy-once",
-                "evaluate the closure only once, even for lists (no closure input)",
-                Some('L'),
+                "eval-once",
+                "evaluate default value closure only once, even for lists (no closure input)",
+                Some('o'),
             )
             .category(Category::Filters)
     }
@@ -54,7 +54,7 @@ impl Command for Default {
         input: PipelineData,
     ) -> Result<PipelineData, ShellError> {
         let empty = call.has_flag(engine_state, stack, "empty")?;
-        let lazy = call.has_flag(engine_state, stack, "lazy")?;
+        let lazy = !call.has_flag(engine_state, stack, "no-eval")?;
         let lazy_once = call.has_flag(engine_state, stack, "lazy-once")?;
         default(engine_state, stack, call, input, empty, lazy, lazy_once)
     }
```

Bahex commented:
Doesn't `upsert` cover most of the `--lazy` cases?

```nushell
> ls | upsert title {|row| $row.name | str upcase }
╭───┬────────────┬──────┬────────┬────────────┬────────────╮
│ # │    name    │ type │  size  │  modified  │   title    │
├───┼────────────┼──────┼────────┼────────────┼────────────┤
│ 0 │ Cargo.toml │ file │ 6.9 kB │ a week ago │ CARGO.TOML │
│ 1 │ LICENSE    │ file │ 1.0 kB │ a week ago │ LICENSE    │
│ 2 │ README.md  │ file │  289 B │ a week ago │ README.MD  │
│ 3 │ src        │ dir  │ 1.0 kB │ a day ago  │ SRC        │
│ 4 │ tests      │ dir  │  192 B │ a week ago │ TESTS      │
╰───┴────────────┴──────┴────────┴────────────┴────────────╯

> [{a:1 b:2} {b:1}] | upsert a {|row| if $in == null { $row.b + 1 } else { $in } }
# or
> [{a:1 b:2} {b:1}] | upsert a {|row| default ($row.b + 1) }
╭───┬───┬───╮
│ # │ a │ b │
├───┼───┼───┤
│ 0 │ 1 │ 2 │
│ 1 │ 2 │ 1 │
╰───┴───┴───╯
```

132ikl commented:
> As an aside, it does seem like we can enable lazy closure evaluation by default, if nobody would mind a breaking change for user scripts.

sounds good to me. I would guess not many people are using a closure lazily, and I think the behavior makes sense.

Mrfiregem commented:
> Doesn't `upsert` cover most of the `--lazy` cases?

For table/record operations, yes. It's basically `upsert` with an `if $in != null` guard, but that's already the case with the current `default` implementation, without this change. The biggest benefits, in my eyes, are

1. Being able to prompt for user input without requiring a whole `if-else` block, if a value is missing (see [this](https://discord.com/channels/601130461678272522/1191485671966261399) Discord thread), which would be useful when writing scripts
2. If calculating a default value would take long time, the previous way of supplying it with a subshell or storing it in a variable would always greedily evaluate, but now it's only run when needed

Bahex commented:
Passing input to `default`'s closure just feels wrong to me, it makes `default` too similar to `upsert`.  

I know it's slightly more verbose but I would rather have `default` as a smaller tool that can be composed with `upsert`.

```nushell
> [{a:1 b:2} {b:1}] | default { $in.b + 1 } a
# vs
> [{a:1 b:2} {b:1}] | upsert a {|row| default { $row.b + 1 } }
```

> This is just my personal preference, I'm not strongly opposed to the PR as it is.

---

Also is `--no-eval` really necessary? With existing commands we need to use closure-in-closure pattern @weirdan pointed out, it would be weird to have a single command that has `--no-eval` while other similar commands don't.
```nushell
.. | insert fn { {..} }
.. | update fn { {..} }

.. | default { {..} } fn
# vs
.. | default --no-eval {..} fn
```

Mrfiregem commented:
So are you suggesting removing the use case of running `default` with a column name and only letting it work on simple values? It sould certainly simplify the command code greatly. For the second point, if we do keep eval-by-default, I don't have a strong opinion one way or another for keeping `--no-eval`, besides that new users might be slightly more confused on how to pass closures (but that can be mentioned in an example)

Mrfiregem commented:
I changed the command to always evaluate closure values, following @weirdan and @Bahex's observations that returning a default value closure is possible without `--no-eval`, and simplified the code such that the default value is always evaluated upfront if it's a closure, which removes the need for both `--no-eval` and `--eval-once`.

If there are no other concerns, I can rewrite the PR description in the initial comment.

weirdan commented:
> the default value is always evaluated upfront if it's a closure

wait, does that mean that `true | default { random int }` evaluates the `{ random int}`? If so, it defeats the purpose of lazy evaluation.

Mrfiregem commented:
Yes, you're right, lol 🤦. I'll fix that right now

weirdan commented:
For the `default {..} col` form, `--once` option still makes sense, I think, to generate the value either on first null/empty, or for every null/empty.

Mrfiregem commented:
Well, I bit the bullet and stopped trying to shove everything in the map function. The result is a large increase in LOC, but the command runs at the same speed for non-closure values (`bench --pretty { ls | default 'N/A' foo } # => 693µs 252ns +/- 63µs 837ns` on my machine), and since closure values are cached for defaulting columns, it's not too slow either (`bench --pretty { ls | default { 'N/A' } foo } # => 2ms 306µs 666ns +/- 292µs 744ns` / `timeit { ls | default { sleep 5sec; 'N/A' } foo } # => 5sec 5ms 644µs 791ns`).

Summary copied from commit message:
> This will only ever run the closure one time, when needed, then cache the values, so long operations are only ever run once, or not at all if not necessary, even for records or tables with multiple columns.
> Another change is allowing multiple column names, and erroring when the user gives a column name when it doesn't make sense for the input, whereas previously the command would always return the pipeline input. This can be reverted if needed.

132ikl commented:
> For the `default {..} col` form, `--once` option still makes sense, I think, to generate the value either on first null/empty, or for every null/empty.

I think the `upsert { default { .. } } col` form suggested by Bahex makes more sense than shoving all of that functionality into `default`

132ikl commented on crates/nu-command/src/filters/default.rs:
```diff
@@ -101,79 +101,219 @@ impl Command for Default {
     let metadata = input.metadata();
-    let value: Value = call.req(engine_state, stack, 0)?;
-    let column: Option<Spanned<String>> = call.opt(engine_state, stack, 1)?;
+    let default_value: Spanned<Value> = call.req(engine_state, stack, 0)?;
```
Values already have an internal span, there's no need to use `Spanned` with `Value`

weirdan commented:
> I think the `upsert { default { .. } } col` form suggested by Bahex makes more sense than shoving all of that functionality into default

Does that imply we want to remove the column argument from `update`? Also, `upsert { default { .. } } col` does not allow evaluating the closure once, and only if required.

132ikl commented:
> Does that imply we want to remove the column argument from `update`? Also, `upsert { default { .. } } col` does not allow evaluating the closure once, and only if required.


Do you mean `default` here instead of `update`? If you mean `default`, then no, it's just that `default` with a column name would have the evaluate once behavior.

It kind of maps like this:

`default --no-eval { echo hi }` -> `default { { echo hi } }`
`default --eval-once { echo hi } column` -> `default { echo hi } column`
`default --eval-multiple { echo hi } column` -> `upsert { default { echo hi } } column` (used `--eval-multiple` here for clarity even though none of the implementations actually had a flag named that, this corresponds to `--lazy` in the original PR)

Mrfiregem commented:
I've updated the original PR description with examples as to how this PR currently functions. Since there's no closure input currently, and in the spirit of `default`, I'm skeptical that `--eval-multiple` is needed. `--no-eval` was also removed in favor of `default { {|| } }` as you both suggested.

132ikl commented on crates/nu-command/src/filters/default.rs:
I think you could simplify a lot of the code by separating out the logic for calculating the default value and the logic for setting that default value. I played around with it a bit, and came up with this enum which encapsulates the logic for the default value calculation. With this, you can construct the `DefaultValue` before doing anything (which does not actually evaluate the value), and any time you need the value you can just call the `DefaultValue::value` method without worrying about whether it's already been calculated or not (if it hasn't been calculated it's calculated on call, and if it has been calculated then it just returns a clone of the value):

```rust
enum DefaultValue<'a> {
    Uncalculated(&'a EngineState, &'a Stack, Spanned<Closure>),
    Calculated(Value),
}

impl<'a> DefaultValue<'a> {
    fn new(engine_state: &'a EngineState, stack: &'a Stack, value: Value) -> Self {
        let span = value.span();
        match value {
            Value::Closure { val, .. } => {
                DefaultValue::Uncalculated(engine_state, stack, (*val).into_spanned(span))
            }
            _ => DefaultValue::Calculated(value),
        }
    }

    fn value(&mut self) -> Result<Value, ShellError> {
        match self {
            DefaultValue::Uncalculated(engine_state, stack, closure) => {
                let closure_eval = ClosureEvalOnce::new(engine_state, stack, closure.item.clone());
                let value = closure_eval
                    .run_with_input(PipelineData::Empty)?
                    .into_value(closure.span)?;
                *self = DefaultValue::Calculated(value.clone());
                Ok(value)
            }
            DefaultValue::Calculated(value) => Ok(value.clone()),
        }
    }
}
```

Let me know what you think :smile:

Mrfiregem replied:
Thank you for this suggestion. After this is done, I'm definitely gonna go through the Rust book from front to back 😅. I probably won't add this today, but this is a much better way than what I came up with.

weirdan commented:
> Do you mean `default` here instead of `update`?

Yes, of course. I must have been distracted at the moment.

> `default` with a column name would have the evaluate once behavior.

Ah, ok, sounds good then.

132ikl commented on crates/nu-command/src/filters/default.rs:
```diff
@@ -1,5 +1,5 @@
-use nu_engine::command_prelude::*;
-use nu_protocol::{ListStream, Signals};
+use nu_engine::{command_prelude::*, ClosureEvalOnce};
+use nu_protocol::{engine::Closure, ListStream, Signals};
```
```suggestion
use nu_engine::{command_prelude::*, ClosureEvalOnce};
use nu_protocol::{engine::Closure, ListStream, Signals};
```

(from diff)

132ikl commented on crates/nu-command/src/filters/default.rs:
```diff
@@ -43,8 +43,18 @@ impl Command for Default {
+            input,
+            default_value,
+            empty,
+            columns.as_slice(),
```
```suggestion
            columns,
```

We can pass ownership of the `Vec` to `default` for usage in the `.map` closure later

132ikl commented on crates/nu-command/src/filters/default.rs:
```diff
@@ -110,70 +139,150 @@ fn default(
     input: PipelineData,
+    default_value: Value,
     default_when_empty: bool,
+    columns: &[String],
```
```suggestion
    columns: Vec<String>,
```
same as above

132ikl commented on crates/nu-command/src/filters/default.rs:
```diff
@@ -110,70 +139,150 @@ fn default(
+    // and set the default value for the specified record columns
+    if !columns.is_empty() {
+        if matches!(input, PipelineData::Value(Value::Record { .. }, _)) {
+            let Value::Record {
+                val: ref mut record,
+                internal_span,
+            } = input.into_value(input_span)?
+            else {
+                unreachable!()
+            };
+            let record = record.to_mut().into_spanned(internal_span);
+            fill_record(record, &mut default_value, columns, default_when_empty)
+                .map(|x| x.into_pipeline_data().set_metadata(metadata))
```
```suggestion
            let record = input.into_value(input_span)?.into_record()?;
            fill_record(
                record,
                input_span,
                &mut default_value,
                columns.as_slice(),
                default_when_empty,
            )
            .map(|x| x.into_pipeline_data_with_metadata(metadata))
```

Since we're already coercing this into a concrete `Value`, we can just extract the `Record` value directly rather than mutating it by reference, which is a bit nicer to work with

132ikl commented on crates/nu-command/src/filters/default.rs:
```diff
@@ -110,70 +139,150 @@ fn default(
+            input,
+            PipelineData::ListStream(..) | PipelineData::Value(Value::List { .. }, _)
+        ) {
+            let mut output_list: Vec<Value> = vec![];
+            for mut item in input {
+                if let Value::Record {
+                    val: ref mut record,
+                    internal_span,
+                } = item
+                {
+                    let record = record.to_mut().into_spanned(internal_span);
+                    item = fill_record(record, &mut default_value, columns, default_when_empty)?;
+                }
+                output_list.push(item);
+            }
+            let ls = ListStream::new(
+                output_list.into_iter(),
+                call.head,
+                engine_state.signals().clone(),
+            );
+            Ok(PipelineData::ListStream(ls, metadata))
```
```suggestion
            // Potential enhancement: add another branch for Value::List,
            // and collect the iterator into a Result<Value::List, ShellError>
            // so we can preemptively return an error for collected lists
            Ok(input
                .into_iter()
                .map(move |item| {
                    let span = item.span();
                    if let Value::Record { val, .. } = item {
                        fill_record(
                            val.into_owned(),
                            span,
                            &mut default_value,
                            columns.as_slice(),
                            default_when_empty,
                        )
                        .unwrap_or_else(|err| Value::error(err, head))
                    } else {
                        item
                    }
                })
                .into_pipeline_data_with_metadata(head, engine_state.signals().clone(), metadata))
```

I think we should try to maintain streaming when possible. While you used a `ListStream` before, the entire stream was implicitly collected by the `for` loop. By using `map` on the `PipelineData` iterator, we can keep the stream intact.

132ikl commented on crates/nu-command/src/filters/default.rs:
```diff
@@ -110,70 +139,150 @@ fn default(
 }
 
+/// A wrapper around the default value to handle closures and caching values
+enum DefaultValue<'a> {
+    Uncalculated(&'a EngineState, &'a Stack, Spanned<Closure>),
+    Calculated(Value),
+}
+
+impl<'a> DefaultValue<'a> {
+    fn new(engine_state: &'a EngineState, stack: &'a Stack, value: Value) -> Self {
+        let span = value.span();
+        match value {
+            Value::Closure { val, .. } => {
+                DefaultValue::Uncalculated(engine_state, stack, (*val).into_spanned(span))
+            }
+            _ => DefaultValue::Calculated(value),
+        }
+    }
+
+    fn value(&mut self) -> Result<Value, ShellError> {
+        match self {
+            DefaultValue::Uncalculated(engine_state, stack, closure) => {
+                let closure_eval = ClosureEvalOnce::new(engine_state, stack, closure.item.clone());
+                let value = closure_eval
+                    .run_with_input(PipelineData::Empty)?
+                    .into_value(closure.span)?;
+                *self = DefaultValue::Calculated(value.clone());
+                Ok(value)
+            }
+            DefaultValue::Calculated(value) => Ok(value.clone()),
+        }
+    }
+
+    fn pipeline_data(&mut self) -> Result<PipelineData, ShellError> {
+        self.value().map(|x| x.into_pipeline_data())
+    }
+}
```
```suggestion
enum DefaultValue {
    Uncalculated(Spanned<ClosureEval>),
    Calculated(Value),
}

impl DefaultValue {
    fn new(engine_state: &EngineState, stack: &Stack, value: Value) -> Self {
        let span = value.span();
        match value {
            Value::Closure { val, .. } => {
                let closure_eval = ClosureEval::new(engine_state, stack, *val);
                DefaultValue::Uncalculated(closure_eval.into_spanned(span))
            }
            _ => DefaultValue::Calculated(value),
        }
    }

    fn value(&mut self) -> Result<Value, ShellError> {
        match self {
            DefaultValue::Uncalculated(closure) => {
                let value = closure
                    .item
                    .run_with_input(PipelineData::Empty)?
                    .into_value(closure.span)?;
                *self = DefaultValue::Calculated(value.clone());
                Ok(value)
            }
            DefaultValue::Calculated(value) => Ok(value.clone()),
        }
    }

    fn pipeline_data(&mut self) -> Result<PipelineData, ShellError> {
        self.value().map(|x| x.into_pipeline_data())
    }
}
```

In order to pass `DefaultValue` into the closure used for `map`, we can't have a reference to EngineState or Stack. Therefore, we have to use `ClosureEval` here, which does unfortunately clone `EngineState` and `Stack`, but I think this is a reasonable tradeoff to preserve streaming (this is how e.g., the `each` command works)

132ikl commented on crates/nu-command/src/filters/default.rs:
```diff
@@ -110,70 +139,150 @@ fn default(
+}
+
+/// Given a record, fill missing columns with a default value
+fn fill_record(
+    record: Spanned<&mut Record>,
+    default_value: &mut DefaultValue,
+    columns: &[String],
+    empty: bool,
+) -> Result<Value, ShellError> {
+    for col in columns {
+        if let Some(val) = record.item.get_mut(col) {
+            if matches!(val, Value::Nothing { .. }) || (empty && val.is_empty()) {
+                *val = default_value.value()?;
+            }
+        } else {
+            record.item.push(col.clone(), default_value.value()?);
+        }
+    }
+    Ok(Value::record(record.item.clone(), record.span))
+}
```
```suggestion
fn fill_record(
    mut record: Record,
    span: Span,
    default_value: &mut DefaultValue,
    columns: &[String],
    empty: bool,
) -> Result<Value, ShellError> {
    for col in columns {
        if let Some(val) = record.get_mut(col) {
            if matches!(val, Value::Nothing { .. }) || (empty && val.is_empty()) {
                *val = default_value.value()?;
            }
        } else {
            record.push(col.clone(), default_value.value()?);
        }
    }
    Ok(Value::record(record, span))
}
```

This just changes the function to operate on the `Record` value directly rather than working with a mutable reference. Previously, `record.item.clone()` is always cloned (so passing by mutable reference isn't really helping). We can avoid that clone by operating on the value directly.

To get a bit into the weeds, we *might* actually clone above when we call `into_record` on the `Value` containing the Record, as created an owned version of `Record` means we might have to pull it out of the `SharedCow`, but a "maybe clone" is better than an "always clone" 😄

132ikl reviewed (changes requested):
Thanks for implementing this! It's looking much better 😄

I did notice that this made it incompatible with streaming when using column names, which is different to the previous behavior. We also are streaming in the no column names `ListStream` branch, so I thought we should try to stream here.

I try to make sure what I'm thinking is actually possible before suggesting it to newer contributors, so I started playing around with it and ended up implementing the whole entire thing before I was able to determine that it is indeed possible. I didn't mean to step on your toes so much in this PR, I don't want you to feel like we don't value your contribution! As you mentioned, this ended up being a surprisingly complicated PR.

I tried to split up my changes into individual GitHub suggestions with explanations so you can understand the changes I made. You should be able to safely apply them as-is. Let me know if you have any questions or anything, and again sorry for accidentally writing so much code on what is ostensibly your PR 😅

Mrfiregem commented:
> I try to make sure what I'm thinking is actually possible before suggesting it to newer contributors, so I started playing around with it and ended up implementing the whole entire thing before I was able to determine that it is indeed possible. I didn't mean to step on your toes so much in this PR, I don't want you to feel like we don't value your contribution! As you mentioned, this ended up being a surprisingly complicated PR.

It's not a problem at all. All of my PRs so far have been things that I wanted to use as a user of Nushell, so I appreciate your help in adding this feature 🙏. I'm definitely learning a lot about Nushell internals, so hopefully I'll be able to tackle other larger issues.

> I tried to split up my changes into individual GitHub suggestions with explanations so you can understand the changes I made. You should be able to safely apply them as-is. Let me know if you have any questions or anything, and again sorry for accidentally writing so much code on what is ostensibly your PR 😅

Thanks for commenting your reasoning behind each change, too. There were some problems because a commit changed tests to use a newer rust edition, but it's all been sorted out easily.

132ikl commented:
thanks so much!! just one last thing, I promise! would you be able to add a couple tests for the new functionality in `nu-command/tests/commands/default.rs`? I meant to mention that yesterday but forgot 😅 

A test to make sure the evaluation is *actually* lazy, a test for lazy returning a closure value, and testing closure evaluation with columns and with `--empty` would be good

132ikl commented:
looks great! thanks so much for your work on this, a long awaited feature! excited to see future contributions from you 😄

NotTheDr01ds commented:
> * To return a closure, the user must wrap the closure they want to return inside another closure, which will be run (`default { $my_closure }`).

This is a breaking change that gave me a startup error based on my Carapace completion closure using a `default` from back in the day when the config record wasn't populated by default.

Wrapping it in `{ $carapace_completer }` works, but definitely needs to be documented as a breaking change since I'm probably not the only one with that config.

NotTheDr01ds commented:
@132ikl The more I look at this, the more I'm not sure about it.  The fact that you have to *wrap* a closure-variable in additional braces to pass it as a value seems (a) to be different than any other place where we pass a closure in a variable, (b) seems too "magic" for this particular case, and (c) changes the pre-PR behavior.

Is it possible to switch the behavior so that:

* An unwrapped variable continues to be treated as it was before, but
* You need to wrap a closure in double `{{ closure }}` to have it lazily evaluated

This would seem more logical to me, as we're *evaluating* the closure in the lazy case.

weirdan commented:
> The fact that you have to wrap a closure-variable in additional braces to pass it as a value seems (a) to be different than any other place where we pass a closure in a variable, (b) seems too "magic" for this particular case,

`insert`/`update`/`upsert` work the same

> * You need to wrap a closure in double {{ closure }} to have it lazily evaluated
> This would seem more logical to me, as we're evaluating the closure in the lazy case.

That's ... too magical and has no parallels in nushell, as far as I can tell.

NotTheDr01ds commented:
Fair enough - Ignore my late night hallucinations :-)

132ikl commented:
just remember to get back to this now, thanks for adding the breaking change tag @NotTheDr01ds, I forgot to add that! not sure if you saw, but I did also open a PR to fix the carapace issue: https://github.com/carapace-sh/carapace-bin/pull/2796
